tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,138 @@
+import com.google.common.collect.ImmutableMap;
+import com.linagora.tmail.james.jmap.settings.JmapSettings;
+import com.linagora.tmail.james.jmap.settings.JmapSettingsRepository;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import scala.collection.JavaConverters;
+
+public class CleanupTrashService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CleanupTrashService.class);
+    public static final JmapSettingsValue WEEKLY_PERIOD = new JmapSettingsValue("weekly");
+    public static final JmapSettingsValue MONTHLY_PERIOD = new JmapSettingsValue("monthly");
+    public static final JmapSettingsKey TRASH_CLEANUP_ENABLED_SETTING = JmapSettings.JmapSettingsKey$.MODULE$.liftOrThrow("trash.cleanup.enabled");
+    public static final JmapSettingsKey TRASH_CLEANUP_PERIOD_SETTING = JmapSettings.JmapSettingsKey$.MODULE$.liftOrThrow("trash.cleanup.period");
+    private static final Map<JmapSettingsValue, Period> mapPeriodSettingToDuration = ImmutableMap.of(WEEKLY_PERIOD,  Period.ofWeeks(1),
+        MONTHLY_PERIOD, Period.ofMonths(1));
+
+    private final UsersRepository usersRepository;
+    private final JmapSettingsRepository jmapSettingsRepository;
+    private final StoreMailboxManager storeMailboxManager;
+    private final SystemMailboxesProvider systemMailboxesProvider;
+    private final Clock clock;
+
+    @Inject
+    public CleanupTrashService(UsersRepository usersRepository, JmapSettingsRepository jmapSettingsRepository, StoreMailboxManager storeMailboxManager,
+                               SystemMailboxesProvider systemMailboxesProvider, Clock clock) {
+        this.usersRepository = usersRepository;
+        this.jmapSettingsRepository = jmapSettingsRepository;
+        this.storeMailboxManager = storeMailboxManager;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash(RunningOptions runningOptions) {
+        return Flux.from(usersRepository.listReactive())
+            .transform(ReactorUtils.<Username, Task.Result>throttle()
+                .elements(runningOptions.getUsersPerSecond())
+                .per(Duration.ofSeconds(1))
+                .forOperation(this::cleanupTrashForSingleUser))
+            .reduce(Task.Result.COMPLETED, Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while accessing users from repository", e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username) {
+        return Mono.from(jmapSettingsRepository.get(username))
+            .flatMap(jmapSettings -> cleanupTrashForSingleUser(username, jmapSettings))
+            .onErrorResume(e -> {
+                LOGGER.error("Error while cleaning trash for user {}", username, e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
```

Comment:
What if a user does not have any JmapSettings (https://github.com/linagora/tmail-backend/blob/master/tmail-backend/jmap/extensions/src/test/scala/com/linagora/tmail/james/jmap/settings/JmapSettingsRepositoryContract.scala#L40)? 
We do not fallback result as `Task.Result.COMPLETED` for that user?

Reply:
:(( IMO it should return an empty map

Reply:
or it should be Optional

Reply:
Maybe, sometimes I am concerned about forgetting to handle `Mono.empty` returned by `jmapSettingsRepository.get` too (could lead to some bugs).

Should we handle return fallback empty JmapSettings object within `jmapSettingsRepository.get` @vttranlina ?

Reply:
Your mean is return `JmapSettings(Map.empty(), JmapSettingsStateFactory.INITIAL)`?
I thought it, But now don't remember why 
___________
IMO, return COMPLETED for the user who hasn't any settings is fine
I remember when grooming, we discussed that users don't have default settings

Reply:
for now, let it be. I will handle null case